modules/ROOT/pages/patterns/shortest-paths.adoc

```diff
@@ -542,18 +542,24 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
 
 [[planning-StatefulShortestPath-operator]]
 == Planning the StatefulShortestPath-operator
+The `StatefulShortestPath(All)` uses a unidirectional breadth first search algorithm, while the `StatefulShortestPath(Into)` uses a bidirectional breadth first search algorithm.
 See xref:planning-and-tuning/operators/operators-detail.adoc#query-plan-stateful-shortest-path-into[the operator-detail page] for a description of the `StatefulShortestPath(All)` and `StatefulShortestPath(Into)` operators.
 
-When the subexpression of the statefulShortestPath-operator matches only one of both boundary nodes, then a `StatefulShortestPath(All)` is planned.
+The planner decides between the two variants using this heuristic:
 
-When the subexpression of the statefulShortestPath-operator matches both boundary nodes, then a `StatefulShortestPath(Into)` is planned only when both boundary nodes match to at most one data node. Otherwise, this plan is avoided when possible.
-Another plan will be favoured where the subexpression matches only one of both boundary nodes followed by an `StatefulShortestPath(All)`.
+* When both boundary nodes have been matched in `MATCH` clause prior to the `MATCH` clause containing the shortest paths query, `StatefulShortestPath(Into)` is planned
+* Otherwise, if both boundary nodes match to at most one data node, then `StatefulShortestPath(Into)` is planned
+* Otherwise, `StatefulShortestPath(All)` is planned
```

Comment:
```suggestion
* Otherwise, if both boundary nodes match to at most one data node, then `StatefulShortestPath(Into)` is planned.
* Otherwise, `StatefulShortestPath(All)` is planned.
```